# Data Transformation
---

In [1]:
import pandas as pd
import numpy as np

# read in our data 
artist = pd.read_csv('./data/artist.csv')
album = pd.read_csv('./data/album.csv')
track = pd.read_csv('./data/track.csv')
track_feat = pd.read_csv('./data/track_feat.csv')

# create function to check if there are duplicate row values in the dfs
def check_for_dups(df, table):
    '''Function takes in a pandas dataframe (df) and the name if the df as a string (table)'''
    potential_dups = []
    # iterate through the columns in the df
    for col in df.columns:
        # check to see if the column values are numbers, if not continue
        if df.dtypes[col] != np.int64 and df.dtypes[col] != np.float64:
            # if 'count' and 'unique' are different then potential duplicate values
            if df[col].describe()['count'] != df[col].describe()['unique']:
                potential_dups.append(col)
    print(f'The following columns in {table} have potential duplicates: {potential_dups}')

# check all dfs for potential duplicate values
check_for_dups(artist, 'artist')
check_for_dups(album, 'album')
check_for_dups(track, 'track')
check_for_dups(track_feat, 'track_feat')

The following columns in artist have potential duplicates: ['genre', 'type']
The following columns in album have potential duplicates: ['album_id', 'album_name', 'external_url', 'image_url', 'release_date', 'type', 'album_uri', 'artist_id']
The following columns in track have potential duplicates: ['track_id', 'song_name', 'external_url', 'explicit', 'type', 'song_uri', 'album_id']
The following columns in track_feat have potential duplicates: ['track_id', 'type', 'song_uri']


## Checking for duplicates
---

## Artist dataframe:

In [2]:
check_for_dups(artist, 'artist')

The following columns in artist have potential duplicates: ['genre', 'type']


## The columns above, genre and type, are good to have duplicate values. Artist df is free of redundant duplicates.

## Album dataframe:

In [3]:
check_for_dups(album, 'album')

The following columns in album have potential duplicates: ['album_id', 'album_name', 'external_url', 'image_url', 'release_date', 'type', 'album_uri', 'artist_id']


## The only columns above that are good to have duplicate values are type, release_date, and artist_id. Let's take a closer look at the album df. 

In [4]:
album[album['album_name'].duplicated(keep=False)]

album_id                   album_name  \
34   3XADnbi4uhYXb7RuSJ7bre    Slime Language 2 (Deluxe)   
35   3ihwKkIMJWmmp1huNH0iWC             Slime Language 2   
36   4z8IoUDFp5dmiQNDlT4hu5    Slime Language 2 (Deluxe)   
37   2MUc9nWkppdNnM8dfz0l1w             Slime Language 2   
66   3aITAVBURujVe8fhI2seeR  Pluto x Baby Pluto (Deluxe)   
67   48xpWR8K6CGpy3ETAym3pt           Pluto x Baby Pluto   
68   27fzM2E0lgovCD7PCq6eh4  Pluto x Baby Pluto (Deluxe)   
69   6HcU64bPPXTHIbWmGblIkT           Pluto x Baby Pluto   
108  3aITAVBURujVe8fhI2seeR  Pluto x Baby Pluto (Deluxe)   
109  48xpWR8K6CGpy3ETAym3pt           Pluto x Baby Pluto   
110  27fzM2E0lgovCD7PCq6eh4  Pluto x Baby Pluto (Deluxe)   
111  6HcU64bPPXTHIbWmGblIkT           Pluto x Baby Pluto   

                                          external_url  \
34   https://open.spotify.com/album/3XADnbi4uhYXb7R...   
35   https://open.spotify.com/album/3ihwKkIMJWmmp1h...   
36   https://open.spotify.com/album/4z8IoUDFp5dmiQN...   
37   https://open.spotify.com/album/2MUc9nWkppdNnM8...   
66   https://open.spotify.com/album/3aITAVBURujVe8f...   
67   https://open.spotify.com/album/48xpWR8K6CGpy3E...   
68   https://open.spotify.com/album/27fzM2E0lgovCD7...   
69   https://open.spotify.com/album/6HcU64bPPXTHIbW...   
108  https://open.spotify.com/album/3aITAVBURujVe8f...   
109  https://open.spotify.com/album/48xpWR8K6CGpy3E...   
110  https://open.spotify.com/album/27fzM2E0lgovCD7...   
111  https://open.spotify.com/album/6HcU64bPPXTHIbW...   

                                             image_url release_date  \
34   https://i.scdn.co/image/ab67616d0000b273bbb071...   2021-04-16   
35   https://i.scdn.co/image/ab67616d0000b2737939ff...   2021-04-16   
36   https://i.scdn.co/image/ab67616d0000b273f376cc...   2021-04-15   
37   https://i.scdn.co/image/ab67616d0000b273f56ecc...   2021-04-15   
66   https://i.scdn.co/image/ab67616d0000b2738efe5b...   2020-11-13   
67   https://i.scdn.co/image/ab67616d0000b27357928e...   2020-11-13   
68   https://i.scdn.co/image/ab67616d0000b27319be1f...   2020-11-12   
69   https://i.scdn.co/image/ab67616d0000b2739e3502...   2020-11-12   
108  https://i.scdn.co/image/ab67616d0000b2738efe5b...   2020-11-13   
109  https://i.scdn.co/image/ab67616d0000b27357928e...   2020-11-13   
110  https://i.scdn.co/image/ab67616d0000b27319be1f...   2020-11-12   
111  https://i.scdn.co/image/ab67616d0000b2739e3502...   2020-11-12   

     total_tracks   type                             album_uri  \
34             31  album  spotify:album:3XADnbi4uhYXb7RuSJ7bre   
35             23  album  spotify:album:3ihwKkIMJWmmp1huNH0iWC   
36             31  album  spotify:album:4z8IoUDFp5dmiQNDlT4hu5   
37             23  album  spotify:album:2MUc9nWkppdNnM8dfz0l1w   
66             24  album  spotify:album:3aITAVBURujVe8fhI2seeR   
67             16  album  spotify:album:48xpWR8K6CGpy3ETAym3pt   
68             24  album  spotify:album:27fzM2E0lgovCD7PCq6eh4   
69             16  album  spotify:album:6HcU64bPPXTHIbWmGblIkT   
108            24  album  spotify:album:3aITAVBURujVe8fhI2seeR   
109            16  album  spotify:album:48xpWR8K6CGpy3ETAym3pt   
110            24  album  spotify:album:27fzM2E0lgovCD7PCq6eh4   
111            16  album  spotify:album:6HcU64bPPXTHIbWmGblIkT   

                  artist_id  
34   2hlmm7s2ICUX0LVIhVFlZQ  
35   2hlmm7s2ICUX0LVIhVFlZQ  
36   2hlmm7s2ICUX0LVIhVFlZQ  
37   2hlmm7s2ICUX0LVIhVFlZQ  
66   4O15NlyKLIASxsJ0PrXPfz  
67   4O15NlyKLIASxsJ0PrXPfz  
68   4O15NlyKLIASxsJ0PrXPfz  
69   4O15NlyKLIASxsJ0PrXPfz  
108  1RyvyyTE3xzB2ZywiAwp0i  
109  1RyvyyTE3xzB2ZywiAwp0i  
110  1RyvyyTE3xzB2ZywiAwp0i  
111  1RyvyyTE3xzB2ZywiAwp0i

### We can see from the above that there are multiple albums with the same name and album id. This is partially due to the fact that 2 artists have a joint album so when data was pulled for each artist the same album showed up twice. When pulling data we might have also pulled the clean and explicit version for these albums.  Let's remove the duplicate albums.

In [5]:
# drop duplicate albums with the same name
album.drop_duplicates(subset=['album_name'], inplace=True)

check_for_dups(album, 'album')

The following columns in album have potential duplicates: ['release_date', 'type', 'artist_id']


## The columns above, release_date, type, and artist_id, are good to have duplicate values. Album df is free of redundant duplicates.

## Track dataframe:

In [6]:
check_for_dups(track, 'track')

The following columns in track have potential duplicates: ['track_id', 'song_name', 'external_url', 'explicit', 'type', 'song_uri', 'album_id']


## The only columns above that are good to have duplicate values are song_name, explicit, type, and album_id . Let's take a closer look at the track df. 

In [7]:
track[track['track_id'].duplicated(keep=False)]

track_id                song_name  \
1121  0anUgBvnA4u5LsHUeLRprc                  Tic Tac   
1122  4V9K7DLPZNhCH77wLsAiNF                My Legacy   
1123  669gYFBhezfYPXQP1hBMOX          Heart In Pieces   
1124  6vbe3GmRG30x2nraEvHudq           Because of You   
1125  2z0LyXYmPgos95olIAQIO3              Bust a Move   
...                      ...                      ...   
1939  5nOh0dj3i48lk2zIG5gp8o  She Never Been To Pluto   
1940  5vpstUHLixzly5KK9I4UPM                  Off Dat   
1941  1IGjaBw48t5slJGHJ6Q1fa   I Don’t Wanna Break Up   
1942  6dtWiXU0WQ5nKGfTAqqJNJ                 Bankroll   
1943  2xiLwSXm0LJC6BCxtbFaCA        Moment of Clarity   

                                           external_url  duration_ms  \
1121  https://open.spotify.com/track/0anUgBvnA4u5LsH...       189147   
1122  https://open.spotify.com/track/4V9K7DLPZNhCH77...       193333   
1123  https://open.spotify.com/track/669gYFBhezfYPXQ...       196834   
1124  https://open.spotify.com/track/6vbe3GmRG30x2nr...       216541   
1125  https://open.spotify.com/track/2z0LyXYmPgos95o...       226285   
...                                                 ...          ...   
1939  https://open.spotify.com/track/5nOh0dj3i48lk2z...       204117   
1940  https://open.spotify.com/track/5vpstUHLixzly5K...       184736   
1941  https://open.spotify.com/track/1IGjaBw48t5slJG...       243936   
1942  https://open.spotify.com/track/6dtWiXU0WQ5nKGf...       186846   
1943  https://open.spotify.com/track/2xiLwSXm0LJC6BC...       222295   

      explicit  disc_number   type                              song_uri  \
1121      True            1  track  spotify:track:0anUgBvnA4u5LsHUeLRprc   
1122      True            1  track  spotify:track:4V9K7DLPZNhCH77wLsAiNF   
1123      True            1  track  spotify:track:669gYFBhezfYPXQP1hBMOX   
1124      True            1  track  spotify:track:6vbe3GmRG30x2nraEvHudq   
1125      True            1  track  spotify:track:2z0LyXYmPgos95olIAQIO3   
...        ...          ...    ...                                   ...   
1939     False            1  track  spotify:track:5nOh0dj3i48lk2zIG5gp8o   
1940     False            1  track  spotify:track:5vpstUHLixzly5KK9I4UPM   
1941     False            1  track  spotify:track:1IGjaBw48t5slJGHJ6Q1fa   
1942     False            1  track  spotify:track:6dtWiXU0WQ5nKGfTAqqJNJ   
1943     False            1  track  spotify:track:2xiLwSXm0LJC6BCxtbFaCA   

                    album_id  
1121  3aITAVBURujVe8fhI2seeR  
1122  3aITAVBURujVe8fhI2seeR  
1123  3aITAVBURujVe8fhI2seeR  
1124  3aITAVBURujVe8fhI2seeR  
1125  3aITAVBURujVe8fhI2seeR  
...                      ...  
1939  6HcU64bPPXTHIbWmGblIkT  
1940  6HcU64bPPXTHIbWmGblIkT  
1941  6HcU64bPPXTHIbWmGblIkT  
1942  6HcU64bPPXTHIbWmGblIkT  
1943  6HcU64bPPXTHIbWmGblIkT  

[160 rows x 9 columns]

### Here we can see that much like how there were duplicate albums, there are duplicate songs in our data. We also have certain albums that have a regular and a deluxe version which is usually the same songs with an additional track or two. Let's remove the duplicate tracks.

In [8]:
# drop duplicate tracks with the same track_id
track.drop_duplicates(subset=['track_id'], inplace=True)

check_for_dups(track, 'track')

The following columns in track have potential duplicates: ['song_name', 'explicit', 'type', 'album_id']


## The columns above, song_name, explicit, type, and album_id, are good to have duplicate values. Track df is free of redundant duplicates.

## Track_feature dataframe:

In [9]:
check_for_dups(track_feat, 'track_feat')

The following columns in track_feat have potential duplicates: ['track_id', 'type', 'song_uri']


## The only column above that is good to have a duplicate value is type Let's take a closer look at the track_feat df. 

In [10]:
track_feat[track_feat['track_id'].duplicated(keep=False)]

track_id  danceability  energy  instrumentalness  \
1121  0anUgBvnA4u5LsHUeLRprc         0.942   0.566               0.0   
1122  4V9K7DLPZNhCH77wLsAiNF         0.890   0.682               0.0   
1123  669gYFBhezfYPXQP1hBMOX         0.872   0.699               0.0   
1124  6vbe3GmRG30x2nraEvHudq         0.781   0.469               0.0   
1125  2z0LyXYmPgos95olIAQIO3         0.804   0.683               0.0   
...                      ...           ...     ...               ...   
1939  5nOh0dj3i48lk2zIG5gp8o         0.834   0.605               0.0   
1940  5vpstUHLixzly5KK9I4UPM         0.895   0.652               0.0   
1941  1IGjaBw48t5slJGHJ6Q1fa         0.621   0.583               0.0   
1942  6dtWiXU0WQ5nKGfTAqqJNJ         0.827   0.723               0.0   
1943  2xiLwSXm0LJC6BCxtbFaCA         0.857   0.639               0.0   

      liveness  loudness  speechiness    tempo            type  valence  \
1121    0.1100    -4.834       0.2260  132.858  audio_features    0.479   
1122    0.1160    -5.249       0.3230  143.979  audio_features    0.526   
1123    0.1030    -6.127       0.0487  138.929  audio_features    0.725   
1124    0.0921    -8.456       0.0549  132.989  audio_features    0.518   
1125    0.0693    -7.344       0.0550  140.035  audio_features    0.229   
...        ...       ...          ...      ...             ...      ...   
1939    0.3070    -5.967       0.2330  132.987  audio_features    0.226   
1940    0.3410    -7.646       0.2890  151.926  audio_features    0.570   
1941    0.1140    -5.642       0.2360  162.174  audio_features    0.534   
1942    0.2110    -5.280       0.2900  139.953  audio_features    0.707   
1943    0.2020    -5.392       0.1390  122.075  audio_features    0.522   

                                  song_uri  
1121  spotify:track:0anUgBvnA4u5LsHUeLRprc  
1122  spotify:track:4V9K7DLPZNhCH77wLsAiNF  
1123  spotify:track:669gYFBhezfYPXQP1hBMOX  
1124  spotify:track:6vbe3GmRG30x2nraEvHudq  
1125  spotify:track:2z0LyXYmPgos95olIAQIO3  
...                                    ...  
1939  spotify:track:5nOh0dj3i48lk2zIG5gp8o  
1940  spotify:track:5vpstUHLixzly5KK9I4UPM  
1941  spotify:track:1IGjaBw48t5slJGHJ6Q1fa  
1942  spotify:track:6dtWiXU0WQ5nKGfTAqqJNJ  
1943  spotify:track:2xiLwSXm0LJC6BCxtbFaCA  

[160 rows x 11 columns]

### Similar to the track dataframe we have the same 160 duplicates due to the similar albums between artists.  Let's remove the duplicate tracks.

In [11]:
# drop duplicate tracks with the same track_id
track_feat.drop_duplicates(subset=['track_id'], inplace=True)

check_for_dups(track_feat, 'track')

The following columns in track have potential duplicates: ['type']


## The type column is good to have duplicates, the track_feat df is free of redundant duplicates.